File: apps/api/app/api/v1/academy.py
"""训练营学习进度 API · B 期刀1(纯增量 · 只读写 academy_progress)。

- POST   /academy/progress/complete   {article_slug}  · CurrentUserDep(强制登录)· 幂等标记学完
- DELETE /academy/progress/complete?article_slug=...   · CurrentUserDep · 取消标记(toggle 用 · 幂等)
- GET    /academy/progress             · OptionalCurrentUserDep(未登录返空进度,不 401)

🔴 红线:学习进度域纯增量 · 不 import / 不碰 交易(virtual/engine)/ 支付 / 会员(subscription/redeem)。
article_slug 用 catalog 校验(只接受已知文章 · 防乱传脏数据)· stage 服务端从目录派生(不信前端)。
"""

from datetime import datetime
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel, Field
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import CurrentUserDep, OptionalCurrentUserDep
from app.core.database import get_db
from app.services.academy.catalog import STAGE_TOTALS, is_valid_slug
from app.services.academy.exam_award import award_membership_if_first_pass
from app.services.academy.exam_results import get_exam_status, record_result
from app.services.academy.exams import (
    PASS_RATIO,
    exam_total,
    has_exam,
    pass_line,
    public_questions,
    score_exam,
)
from app.services.academy.progress import get_progress, mark_complete, unmark_complete

router = APIRouter(prefix="/academy", tags=["academy"])

DbDep = Annotated[AsyncSession, Depends(get_db)]

_TOTAL_ARTICLES = sum(STAGE_TOTALS.values())


class CompleteIn(BaseModel):
    article_slug: str = Field(min_length=1, max_length=32)


class CompleteOut(BaseModel):
    article_slug: str
    stage: str
    completed_at: datetime
    newly_completed: bool  # True=本次新标 / False=之前已标(幂等)


class UncompleteOut(BaseModel):
    article_slug: str
    removed: bool  # True=删了一条 / False=本就没标(幂等)


class ProgressOut(BaseModel):
    completed_slugs: list[str]          # 已完成的文章 slug(前端据此画绿勾)
    by_stage: dict[str, int]            # 各阶已完成数(进度 X)
    stage_totals: dict[str, int]        # 各阶文章总数(进度 Y · 后端目录权威)
    total_completed: int
    total_articles: int


def _require_valid_slug(raw: str) -> str:
    slug = raw.strip()
    if not is_valid_slug(slug):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"未知文章 slug: {slug}",
        )
    return slug


@router.post("/progress/complete", response_model=CompleteOut)
async def complete_article(
    payload: CompleteIn, user: CurrentUserDep, db: DbDep,
) -> CompleteOut:
    """标记学完(幂等)· 强制登录。slug 非法 → 400。"""
    slug = _require_valid_slug(payload.article_slug)
    row, newly = await mark_complete(db, user_id=user.id, article_slug=slug)
    return CompleteOut(
        article_slug=row.article_slug,
        stage=row.stage,
        completed_at=row.completed_at,
        newly_completed=newly,
    )


@router.delete("/progress/complete", response_model=UncompleteOut)
async def uncomplete_article(
    user: CurrentUserDep,
    db: DbDep,
    article_slug: Annotated[str, Query(min_length=1, max_length=32)],
) -> UncompleteOut:
    """取消标记(幂等 · toggle 用)· 强制登录。slug 非法 → 400。"""
    slug = _require_valid_slug(article_slug)
    removed = await unmark_complete(db, user_id=user.id, article_slug=slug)
    return UncompleteOut(article_slug=slug, removed=removed)


@router.get("/progress", response_model=ProgressOut)
async def get_my_progress(user: OptionalCurrentUserDep, db: DbDep) -> ProgressOut:
    """当前用户学习进度 · 未登录返回空进度(不 401 · 不破坏游客浏览)。"""
    if user is None:
        return ProgressOut(
            completed_slugs=[], by_stage={}, stage_totals=dict(STAGE_TOTALS),
            total_completed=0, total_articles=_TOTAL_ARTICLES,
        )
    slugs, by_stage = await get_progress(db, user_id=user.id)
    return ProgressOut(
        completed_slugs=slugs,
        by_stage=by_stage,
        stage_totals=dict(STAGE_TOTALS),
        total_completed=len(slugs),
        total_articles=_TOTAL_ARTICLES,
    )


# ===== 模块结业测验(刀2)· ★防作弊:答案只在后端,前端拿题不拿答案、提交选项后端判分 =====


class ExamQuestionPublic(BaseModel):
    stem: str
    options: list[str]  # ★ 不含 answer_index(防作弊)


class ExamQuestionsOut(BaseModel):
    stage: str
    questions: list[ExamQuestionPublic]
    total: int
    pass_line: int   # 及格题数(≥ 即达标)
    pass_ratio: float


class SubmitExamIn(BaseModel):
    stage: str = Field(min_length=1, max_length=16)
    # 每题选中的【原始】选项下标(前端洗牌后须映射回原序提交)· 缺答/越界按答错
    answers: list[int] = Field(default_factory=list)


class QuestionResultOut(BaseModel):
    question_index: int
    your_answer: int | None
    correct_answer: int       # 提交后才回传(供复盘)
    is_correct: bool
    explanation: str


class SubmitExamOut(BaseModel):
    stage: str
    score: int
    total: int
    pass_line: int
    passed: bool
    results: list[QuestionResultOut]
    # 刀3:首次达标发 1 周会员 · membership_awarded=True 仅首次达标(重考不重复发)
    membership_awarded: bool = False
    new_expires_at: datetime | None = None  # 本次发了会员才有(新会员到期日)


class ExamStatusItem(BaseModel):
    stage: str
    passed: bool       # 曾达标(任一次)
    best_score: int
    total: int
    attempts: int


class ExamResultsOut(BaseModel):
    results: list[ExamStatusItem]


def _require_valid_stage(raw: str) -> str:
    stage = raw.strip()
    if not has_exam(stage):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"未知模块或暂无结业测验: {stage}",
        )
    return stage


@router.get("/exam", response_model=ExamQuestionsOut)
async def get_exam(stage: Annotated[str, Query(min_length=1, max_length=16)]) -> ExamQuestionsOut:
    """取某模块结业测验题 · ★只下发题干+选项(无答案)· 公开可读(答题需登录)。"""
    s = _require_valid_stage(stage)
    total = exam_total(s)
    return ExamQuestionsOut(
        stage=s,
        questions=[
            ExamQuestionPublic(stem=q.stem, options=list(q.options))
            for q in public_questions(s)
        ],
        total=total,
        pass_line=pass_line(total),
        pass_ratio=PASS_RATIO,
    )


@router.post("/exam/submit", response_model=SubmitExamOut)
async def submit_exam(
    payload: SubmitExamIn, user: CurrentUserDep, db: DbDep,
) -> SubmitExamOut:
    """提交结业测验 · ★后端用 exams.py 答案重新判分(前端传的分数一律不信)· 记成绩(可重考)。"""
    s = _require_valid_stage(payload.stage)
    scored = score_exam(s, payload.answers)
    await record_result(
        db, user_id=user.id, stage=s,
        score=scored.score, total=scored.total, passed=scored.passed,
    )
    # 刀3:达标 → 首次发 1 周会员(★复用 extend_subscription · UNIQUE 原子幂等只发一次)·
    #   未达标不发;重考已达标不重复发。只碰发会员一处,不碰交易/支付。
    membership_awarded = False
    new_expires_at = None
    if scored.passed:
        membership_awarded, new_expires_at = await award_membership_if_first_pass(
            db, user_id=user.id, stage=s,
        )
    return SubmitExamOut(
        stage=s,
        score=scored.score,
        total=scored.total,
        pass_line=scored.pass_line,
        passed=scored.passed,
        results=[
            QuestionResultOut(
                question_index=r.question_index,
                your_answer=r.your_answer,
                correct_answer=r.correct_answer,
                is_correct=r.is_correct,
                explanation=r.explanation,
            )
            for r in scored.results
        ],
        membership_awarded=membership_awarded,
        new_expires_at=new_expires_at,
    )


@router.get("/exam/results", response_model=ExamResultsOut)
async def get_exam_results(user: OptionalCurrentUserDep, db: DbDep) -> ExamResultsOut:
    """当前用户各模块结业状态 · 未登录返回空(不 401)。"""
    if user is None:
        return ExamResultsOut(results=[])
    status_list = await get_exam_status(db, user_id=user.id)
    return ExamResultsOut(
        results=[
            ExamStatusItem(
                stage=s.stage,
                passed=s.passed,
                best_score=s.best_score,
                total=s.total,
                attempts=s.attempts,
            )
            for s in status_list
        ],
    )

File: apps/cloudflare-api/src/academy.ts
import examBankJson from '../../api/app/services/academy/exam_questions.json'
import examBankEnJson from '../../api/app/services/academy/exam_questions.en.json'
import {
  ACADEMY_ARTICLES,
  ACADEMY_STAGES,
} from '../../web/content/academy/manifest'

import { authenticate, type AuthenticatedUser } from './auth'
import { COMMERCIAL_MEMBERSHIP_ENABLED } from './features'
import {
  HttpError,
  bearerToken,
  jsonResponse,
  readJsonObject,
  requireString,
} from './http'

const PASS_RATIO = 0.8
const AWARD_DAYS = 7
const DAY_MS = 24 * 60 * 60 * 1_000

type ExamQuestion = Readonly<{
  question: string
  options: string[]
  answerIndex: number
  explanation: string
}>

type ExamBank = Readonly<Record<string, ExamQuestion[]>>

const EXAM_BANK = examBankJson as ExamBank
const EXAM_BANK_EN = examBankEnJson as ExamBank
const ARTICLE_STAGE = new Map(
  ACADEMY_ARTICLES.map((article) => [article.slug, article.stage]),
)
const STAGE_TOTALS = Object.fromEntries(
  ACADEMY_STAGES.map((stage) => [
    stage.slug,
    ACADEMY_ARTICLES.filter((article) => article.stage === stage.slug).length,
  ]),
)
const TOTAL_ARTICLES = ACADEMY_ARTICLES.length

async function optionalAuthenticate(
  request: Request,
  env: Env,
): Promise<AuthenticatedUser | null> {
  return bearerToken(request) ? authenticate(request, env) : null
}

function validArticleSlug(value: unknown): {
  slug: string
  stage: string
} {
  if (typeof value !== 'string' || value.length < 1 || value.length > 32) {
    throw new HttpError(422, 'article_slug 格式无效')
  }
  const slug = value.trim()
  const stage = ARTICLE_STAGE.get(slug)
  if (!stage) throw new HttpError(400, `未知训练营文章：${slug}`)
  return { slug, stage }
}

function validStage(value: string, locale: 'zh' | 'en' = 'zh'): {
  stage: string
  questions: ExamQuestion[]
} {
  const stage = value.trim()
  const questions = (locale === 'en' ? EXAM_BANK_EN : EXAM_BANK)[stage]
  if (!questions) throw new HttpError(400, `未知结业测验模块：${stage}`)
  return { stage, questions }
}

function emptyProgress() {
  return {
    completed_slugs: [] as string[],
    by_stage: {} as Record<string, number>,
    stage_totals: STAGE_TOTALS,
    total_completed: 0,
    total_articles: TOTAL_ARTICLES,
  }
}

async function getProgress(
  request: Request,
  env: Env,
  requestId: string,
): Promise<Response> {
  const auth = await optionalAuthenticate(request, env)
  if (!auth) {
    return jsonResponse(emptyProgress(), 200, requestId, request.method)
  }
  const rows = await env.DB
    .prepare(
      `SELECT article_slug, stage
       FROM academy_progress
       WHERE user_id = ?
       ORDER BY completed_at ASC`,
    )
    .bind(auth.user.id)
    .all<{ article_slug: string; stage: string }>()
  const byStage: Record<string, number> = {}
  const completedSlugs = rows.results.map((row) => {
    byStage[row.stage] = (byStage[row.stage] ?? 0) + 1
    return row.article_slug
  })
  return jsonResponse(
    {
      completed_slugs: completedSlugs,
      by_stage: byStage,
      stage_totals: STAGE_TOTALS,
      total_completed: completedSlugs.length,
      total_articles: TOTAL_ARTICLES,
    },
    200,
    requestId,
    request.method,
  )
}

async function completeArticle(
  request: Request,
  env: Env,
  requestId: string,
): Promise<Response> {
  const { user } = await authenticate(request, env)
  const body = await readJsonObject(request)
  const { slug, stage } = validArticleSlug(body.article_slug)
  const completedAt = Date.now()
  const inserted = await env.DB
    .prepare(
      `INSERT OR IGNORE INTO academy_progress
        (user_id, article_slug, stage, completed_at)
       VALUES (?, ?, ?, ?)`,
    )
    .bind(user.id, slug, stage, completedAt)
    .run()
  const row = await env.DB
    .prepare(
      `SELECT completed_at
       FROM academy_progress
       WHERE user_id = ? AND article_slug = ?`,
    )
    .bind(user.id, slug)
    .first<{ completed_at: number }>()
  if (!row) throw new Error('academy progress insert did not persist')
  return jsonResponse(
    {
      article_slug: slug,
      stage,
      completed_at: new Date(row.completed_at).toISOString(),
      newly_completed: inserted.meta.changes === 1,
    },
    200,
    requestId,
    request.method,
  )
}

async function uncompleteArticle(
  request: Request,
  env: Env,
  requestId: string,
): Promise<Response> {
  const { user } = await authenticate(request, env)
  const { slug } = validArticleSlug(
    new URL(request.url).searchParams.get('article_slug'),
  )
  const deleted = await env.DB
    .prepare(
      `DELETE FROM academy_progress
       WHERE user_id = ? AND article_slug = ?`,
    )
    .bind(user.id, slug)
    .run()
  return jsonResponse(
    { article_slug: slug, removed: deleted.meta.changes === 1 },
    200,
    requestId,
    request.method,
  )
}

function examQuestions(
  request: Request,
  requestId: string,
): Response {
  const stageParam = new URL(request.url).searchParams.get('stage')
  const locale = new URL(request.url).searchParams.get('locale') === 'en' ? 'en' : 'zh'
  if (!stageParam || stageParam.length > 16) {
    throw new HttpError(422, 'stage 格式无效')
  }
  const { stage, questions } = validStage(stageParam, locale)
  const total = questions.length
  return jsonResponse(
    {
      stage,
      questions: questions.map(({ question, options }) => ({
        stem: question,
        options,
      })),
      total,
      pass_line: Math.ceil(total * PASS_RATIO),
      pass_ratio: PASS_RATIO,
    },
    200,
    requestId,
    request.method,
  )
}

function parseAnswers(value: unknown): number[] {
  if (!Array.isArray(value) || value.length > 100) {
    throw new HttpError(422, 'answers 格式无效')
  }
  if (
    value.some(
      (answer) => typeof answer !== 'number' || !Number.isSafeInteger(answer),
    )
  ) {
    throw new HttpError(422, 'answers 必须是整数数组')
  }
  return value as number[]
}

async function submitExam(
  request: Request,
  env: Env,
  requestId: string,
): Promise<Response> {
  const { user } = await authenticate(request, env)
  const body = await readJsonObject(request)
  const locale = body.locale === 'en' ? 'en' : 'zh'
  const { stage, questions } = validStage(
    requireString(body, 'stage', { min: 1, max: 16 }),
    locale,
  )
  const answers = parseAnswers(body.answers)
  let score = 0
  const results = questions.map((question, questionIndex) => {
    const yourAnswer = answers[questionIndex] ?? null
    const isCorrect = yourAnswer === question.answerIndex
    if (isCorrect) score += 1
    return {
      question_index: questionIndex,
      your_answer: yourAnswer,
      correct_answer: question.answerIndex,
      is_correct: isCorrect,
      explanation: question.explanation,
    }
  })
  const total = questions.length
  const passLine = Math.ceil(total * PASS_RATIO)
  const passed = total > 0 && score >= passLine
  const timestamp = Date.now()
  let membershipAwarded = false
  let newExpiresAt: string | null = null

  if (passed && COMMERCIAL_MEMBERSHIP_ENABLED) {
    const awardId = crypto.randomUUID()
    const statements = await env.DB.batch([
      env.DB
        .prepare(
          `INSERT INTO academy_exam_results
            (id, user_id, stage, score, total, passed, created_at)
           VALUES (?, ?, ?, ?, ?, 1, ?)`,
        )
        .bind(
          crypto.randomUUID(),
          user.id,
          stage,
          score,
          total,
          timestamp,
        ),
      env.DB
        .prepare(
          `INSERT OR IGNORE INTO academy_exam_awards
            (id, user_id, stage, awarded_days, awarded_at)
           VALUES (?, ?, ?, ?, ?)`,
        )
        .bind(awardId, user.id, stage, AWARD_DAYS, timestamp),
      env.DB
        .prepare(
          `UPDATE users
           SET subscription_expires_at =
                 CASE
                   WHEN subscription_expires_at IS NULL
                     OR subscription_expires_at < ?
                   THEN ?
                   ELSE subscription_expires_at + ?
                 END,
               updated_at = ?
           WHERE id = ?
             AND EXISTS (
               SELECT 1 FROM academy_exam_awards WHERE id = ?
             )`,
        )
        .bind(
          timestamp,
          timestamp + AWARD_DAYS * DAY_MS,
          AWARD_DAYS * DAY_MS,
          timestamp,
          user.id,
          awardId,
        ),
    ])
    membershipAwarded = statements[1]?.meta.changes === 1
    if (membershipAwarded) {
      const expiry = await env.DB
        .prepare('SELECT subscription_expires_at FROM users WHERE id = ?')
        .bind(user.id)
        .first<{ subscription_expires_at: number }>()
      newExpiresAt = expiry
        ? new Date(expiry.subscription_expires_at).toISOString()
        : null
    }
  } else {
    await env.DB
      .prepare(
        `INSERT INTO academy_exam_results
          (id, user_id, stage, score, total, passed, created_at)
         VALUES (?, ?, ?, ?, ?, ?, ?)`,
      )
      .bind(
        crypto.randomUUID(),
        user.id,
        stage,
        score,
        total,
        passed ? 1 : 0,
        timestamp,
      )
      .run()
  }

  return jsonResponse(
    {
      stage,
      score,
      total,
      pass_line: passLine,
      passed,
      results,
      membership_awarded: membershipAwarded,
      new_expires_at: newExpiresAt,
    },
    200,
    requestId,
    request.method,
  )
}

async function examResults(
  request: Request,
  env: Env,
  requestId: string,
): Promise<Response> {
  const auth = await optionalAuthenticate(request, env)
  if (!auth) {
    return jsonResponse({ results: [] }, 200, requestId, request.method)
  }
  const rows = await env.DB
    .prepare(
      `SELECT
         stage,
         MAX(passed) AS passed,
         MAX(score) AS best_score,
         MAX(total) AS total,
         COUNT(*) AS attempts
       FROM academy_exam_results
       WHERE user_id = ?
       GROUP BY stage
       ORDER BY MIN(created_at) ASC`,
    )
    .bind(auth.user.id)
    .all<{
      stage: string
      passed: number
      best_score: number
      total: number
      attempts: number
    }>()
  return jsonResponse(
    {
      results: rows.results.map((row) => ({
        stage: row.stage,
        passed: row.passed === 1,
        best_score: row.best_score,
        total: row.total,
        attempts: row.attempts,
      })),
    },
    200,
    requestId,
    request.method,
  )
}

export async function handleAcademyRoute(
  request: Request,
  env: Env,
  requestId: string,
): Promise<Response | null> {
  const path = new URL(request.url).pathname
  const route = `${request.method} ${path}`
  switch (route) {
    case 'GET /api/v1/academy/progress':
      return getProgress(request, env, requestId)
    case 'POST /api/v1/academy/progress/complete':
      return completeArticle(request, env, requestId)
    case 'DELETE /api/v1/academy/progress/complete':
      return uncompleteArticle(request, env, requestId)
    case 'GET /api/v1/academy/exam':
      return examQuestions(request, requestId)
    case 'POST /api/v1/academy/exam/submit':
      return submitExam(request, env, requestId)
    case 'GET /api/v1/academy/exam/results':
      return examResults(request, env, requestId)
    default:
      return path.startsWith('/api/v1/academy/')
        ? jsonResponse(
            { detail: 'Route not found' },
            404,
            requestId,
            request.method,
          )
        : null
  }
}
